### ISSUE TITLE
[extension] Maintenance1 compatibility

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#639]
The Maintenance1 appendix discusses that `VK_FORMAT_FEATURE_TRANSFER_*_BIT_KHR` was added to allow indicate support for formats that can do transfer **only**.

But it is not clear, that any other format feature implies that `VK_FORMAT_FEATURE_TRANSFER_*_BIT_KHR` should be set.

And some vendors already seem to publish drivers that support a format without transfer, which seems invalid to me.

Also:

> Note:
If no format feature flags are supported, the format itself is not supported, and images of that format cannot be created.

I think this is important and should not be a Note.
Also it is not clear from this if `0` is always returned for unsupported formats, or if enumerants for non-enabled extensions may be returned

**Comment by NicolBolas** [comment]
> I think this is important and should not be a Note.

VkImage's Valid Usage rules specify that the image's `usage` field must match the `format`'s features field. So if none of the bits in the format's feature set pertain to a valid `usage` bit in the image, you cannot create an image of that format.

That's why it's a note: it simply restate's explicitly what has already been said implicitly.

**Comment by krOoze** [comment]
@NicolBolas Right. So this again boils down to Issue #533 a bit. The unextended spec says bit of a different story in the Note.

Still, should not be a Note. *Doing the redundancy check:) With the Note ommited you would not know if you can create `TRANSFER_USAGE` image with `features==0` or not.

**Comment by janharaldfredriksen-arm** [comment]
Also note that the required formats are covered by:

> Formats that are required to support VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT must also support VK_FORMAT_FEATURE_TRANSFER_SRC_BIT_KHR and VK_FORMAT_FEATURE_TRANSFER_DST_BIT_KHR.

Prior to maintenance1, you couldn't express transfer-only, so if the format was supported at all it must have had at least one of the other feature bits set. With the additional transfer bits, implementations can potentially expose additional formats.

I don't see how the spec is not backwards compatible here?

**Comment by krOoze** [comment]
@janharaldfredriksen-arm Like so:

core Vulkan implementation exposes: `FORMAT_R8G8B8` with `FEATURE_COLOR_ATTACHMENT`. And so implicitly `TRANSFER` feature.

extended Vulkan implementation  exposes: `FORMAT_R8G8B8` with `FEATURE_COLOR_ATTACHMENT` without the `FEATURE_TRANSFER` flags.

Non-extension aware app uses `TRANSFER`. It is valid and runds fine on core Vulkan. It fails on extended Vulkan, because it is forced into invalid usage it does not know of.

**Comment by NicolBolas** [comment]
> extended Vulkan implementation exposes: FORMAT_R8G8B8 with FEATURE_COLOR_ATTACHMENT without the FEATURE_TRANSFER flags.

This is where the failure happened.

The compatibility rule is _not_ that implementations will do the exact same thing unless you explicitly change how you're doing something. It's that you'll get the same _behavior_.

To preserve this, Vulkan implementations that exposed any formats without `maintenance1` ought to expose the same formats with the transfer features when `maintenance1` is active. It can expose additional, transfer-only formats.

The hypothetical implementation here did not do this correctly.

As for the other issue (how a core-only application should behave when run on a `maintenance1` implementation), I would personally say that this is user error. Extensions in Vulkan are *opt-in*, not like OpenGL where you're forced into them. By an application declaring that they're using `maintenance1`, they are explicitly declaring that they're following that amended specification.

It's up to the Vulkan implementation to respond correctly to the required extensions. It's up to the Vulkan *application* to ask for the right extensions and abide by those rules. If you ask for it, then you should be ready to get it.

I know that's not what the Guidelines say about extension design; the stuff I found in #638. But I think that's how it ought to work. Vulkan has the chance to use this kind of design because its extensions are opt-in, rather than opt-out.

**Comment by krOoze** [comment]
> The hypothetical implementation here did not do this correctly.

Well yeah. But unless #638 is included in the spec, then there is nothing to forbid it that I could find. Note might be nice too; easier mistakes were made in existing drivers...

> If you ask for it, then you should be ready to get it.

That may make maintenance pretty hard. Your library/app would have to be ready for all 2<sup>N</sup> combinations of enabled extensions.

> Vulkan has the chance to use this kind of design because its extensions are opt-in, rather than opt-out.

Though that would make Extensions bit of a duplicate of Features. I mean if extension needs to mess default behaviors (let's say disable mandatory formats) it can introduce a new Feature anyway to achieve it.

**Comment by NicolBolas** [comment]
>  Your library/app would have to be ready for all 2<sup>N</sup> combinations of enabled extensions.

You only have to be ready for the particular combination of extensions *you enable*. Or more specifically, you have to be ready for each extension you enable.

If you enable `maintenance1`, you have to be ready for how it interacts with formats that you're using for transfer operations. If you don't enable it, you don't have to be ready for it.

With the odd exception for extensions which have well-defined interactions, I don't see where this "combination" problem exists.

> There also seems to be no cost for having the current compatibility model, while giving some nice guarantees.

There is a cost: under such a model, extensions cannot alter functionality. Which makes it impossible for an extensions to say what `maintenance1` says currently: that with this extension, you also have to check some other things to see if this thing you could formerly do will still work.

> Though that would make Extensions bit of a duplicate of Features. I mean if extension needs to mess default behaviors (let's say disable mandatory formats) it can introduce a new Feature anyway to achieve it.

I don't see the point of that. Extensions are _already_ opt-in; if you don't ask for an extension, the implementation isn't allowed to just give it to you anyway.

If you look at the extensions that add new features to Vulkan, you see that they are for fine-grained control over aspects of the feature. So rather than having 4 separate 16-bit extensions for each of the 4 places where you can use 16-bit values, you have one 16-bit extension and 4 extension features to say where the implementation supports their use.

I see no point in having an extension add a feature to turn on the extension's capabilities, when turning on the extension _already_ requires an explicit user requests. If the purpose of an extension is to "disable mandatory formats", why should the user who asks for that extension also have to enable a feature?

**Comment by krOoze** [comment]
> You only have to be ready for the particular combination of extensions you enable. Or more specifically, you have to be ready for each extension you enable.

And any extension anyone can enable in the future. The way I understood your proposition, it is simply maintenance hazard; and I don't see any benefit making it worth the cost. If you want extension to make more evil stuff, you can achieve it all the same, it is only bit uglier (similar rationale as for `adfhadfhdfh_cast<adhfadfh>()` in C++).

We are getting off-topic though. Maybe make an Enhancement Issue, if you think it is worth consideration.

**Comment by NicolBolas** [comment]
> And any extension anyone can enable in the future.

"Anyone" cannot "enable" extensions for you. Only your application can enable extensions.

> Maybe make an Enhancement Issue, if you think it is worth consideration.

You say "worth consideration" as though it's not already happening. Because as you yourself have pointed out, despite Khronos's guidelines they're already doing it.

The question is what should be changed: the guideline or the extensions that violate it? I say it should be the guideline.

**Comment by krOoze** [comment]
> "Anyone" cannot "enable" extensions for you. Only your application can enable extensions. I fail to see the maintenance hazard.

You enable an extension because you want to try the feature in a localized manner. All of your preexisting 2 MLOC stops working because you removed that kind of compatibility and extension makers do whatever they please (essentially making many forks of Vulkan that differ in tiny details).

> The question is what should be changed: the guideline or the extensions that violate it? I say it should be the guideline.

In this case, in respect to the feature flags, IMO the extension.
In general case, I dunno. Seems like a bad idea at first glance and I did not want to get into that in this thread.

**Comment by janharaldfredriksen-arm** [comment]
@krOoze I don't think it does my quote implies that. Note that FORMAT_FEATURE_TRANSFER_SRC/DST_BIT is only described as being valid for images (i.e. linearTilingFeatures and optimalTilingFeatures). I'm reading this as saying that buffers can always be transferred.

**Comment by krOoze** [comment]
@janharaldfredriksen-arm Well, let's see:

In 1.0, **any** feature flags set implies transfer usage being supported.

Based on Table 47. Mandatory format support: 16-bit channels, `VK_FORMAT_R16_UNORM` is required to support `VK_FORMAT_FEATURE_UNIFORM_TEXEL_BUFFER_BIT`, therefore it also supports transfers.

In 1.1 `VK_FORMAT_FEATURE_TRANSFER_X_BIT` was added, and also the language:

> Formats that are required to support `VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT` **must** also support `VK_FORMAT_FEATURE_TRANSFER_SRC_BIT_KHR` and `VK_FORMAT_FEATURE_TRANSFER_DST_BIT_KHR`.

Which, unless otherwisely said, implies that formats that are not required to support `VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT` **may** not also support `VK_FORMAT_FEATURE_TRANSFER_SRC_BIT_KHR` and `VK_FORMAT_FEATURE_TRANSFER_DST_BIT_KHR`.

Based on Table 47. Mandatory format support: 16-bit channels, `VK_FORMAT_R16_UNORM` is **not** required to support `VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT`, therefore it does **not** have to support transfers.

Application that creates `VK_FORMAT_R16_UNORM` resource (irregardless of being buffer or image) using mandatory support and minimal limits is valid in 1.0. That same application triggers undefined behavior in 1.1, which implies 1.1 is not backwards compatible with 1.0.
